Xamarin.Forms.Platform.Android/CarouselView/CarouselViewRenderer.cs

```diff
@@ -0,0 +1,393 @@
+		// To avoid triggering Position changed
+		bool isSwiping;
+
+		void ViewPager_PageSelected (object sender, ViewPager.PageSelectedEventArgs e)
+		{
+			isSwiping = true;
+			Element.Position = e.Position;
+			isSwiping = false;
+		}
+
+		void ViewPager_PageScrollStateChanged (object sender, ViewPager.PageScrollStateChangedEventArgs e)
+		{
+			if (e.State == 0 && !isSwiping) {
+				Element.PositionSelected?.Invoke(Element, EventArgs.Empty);
+			}
+		}
+
+		void configPosition()
+		{
+			isSwiping = true;
+			if (Element.ItemsSource != null)
+			{
+				if (Element.Position > Element.ItemsSource.Count - 1)
+					Element.Position = Element.ItemsSource.Count - 1;
+
+				if (Element.Position == -1)
+					Element.Position = 0;
+			}
+			else {
+				Element.Position = 0;
+			}
+			isSwiping = false;
+
+			indicator.mSnapPage = Element.Position;
+		}
+
+		public async Task InsertItem(object item, int position)
+		{
+			if (Element != null && viewPager != null && Element.ItemsSource != null)
+			{
+				if (position > Element.ItemsSource.Count)
+					throw new CarouselViewException("Page cannot be inserted at a position bigger than ItemsSource.Count");
+
+				if (position == -1)
+					Element.ItemsSource.Add(item);
+				else
+					Element.ItemsSource.Insert(position, item);
+
+				if (position == Element.Position)
+				{
+					viewPager.Adapter = new PageAdapter(Element);
+					viewPager.SetCurrentItem(Element.Position, false);
+				}
+				else
+					viewPager.Adapter.NotifyDataSetChanged();
+
+				await Task.Delay(100);
+
+				if (Element.ItemsSource.Count == 1)
+					Element.PositionSelected?.Invoke(Element, EventArgs.Empty);
+			}
+		}
+
+		// Android ViewPager is the most complicated piece of code ever :)
+		public async Task RemoveItem(int position)
+		{
+			if (Element != null && viewPager != null && Element.ItemsSource != null && Element.ItemsSource?.Count > 0) {
+				
+				isSwiping = true;
+
+				if (position > Element.ItemsSource.Count - 1)
+					throw new CarouselViewException("Page cannot be removed at a position bigger than ItemsSource.Count - 1");
+
+				if (Element.ItemsSource?.Count == 1)
+				{
+					Element.ItemsSource.RemoveAt(position);
+					viewPager.Adapter = new PageAdapter(Element);
+					viewPager.SetCurrentItem(Element.Position, false);
+
+					indicator.SetViewPager(viewPager);
+				}
+				else {
+
+					if (position == Element.Position)
+					{
+
+						var newPos = position - 1;
+						if (newPos == -1)
+							newPos = 0;
+
+						if (position == 0)
+						{
+
+							viewPager.SetCurrentItem(1, Element.AnimateTransition);
+
+							await Task.Delay(100);
+
+							Element.ItemsSource.RemoveAt(position);
+
+							//viewPager.Adapter = new PageAdapter(Element, viewPager);
+							viewPager.Adapter.NotifyDataSetChanged();
+							viewPager.SetCurrentItem(0, false);
+
+							Element.Position = 0;
+
+						}
+						else {
+
+							viewPager.SetCurrentItem(newPos, Element.AnimateTransition);
+
+							await Task.Delay(100);
```

Comment:
@alexrainman: In my testing of your control, these `Delays` were unnecessary and made the whole app hang for many seconds under load testing (i.e., adding thousands of Views to the CarouselView). I removed them and noticed no side effects.